### ISSUE TITLE
Redragon K556 Devarajas

### ISSUE THREAD

**Issue by CalcProgrammer1** [smp4488/dk63#2]
I've been reverse engineering keyboard protocols for my OpenRGB project.  I picked up a Redragon K556 and reversed its protocol, but was unable to find a mode that allows for direct control of the keyboard LEDs over USB so I decided to start looking at the hardware.  I took it apart and snapped some pics of the insides.  It seems to use the same microcontroller as yours.  I'm really excited to see that someone else has looked into this!  I want to help in this effort.  I'm an embedded software engineer and am familiar with a lot of reverse engineering stuff.  Haven't read through your documentation yet but I am definitely going to soon.

My keyboard micro:

![20200727_223519](https://user-images.githubusercontent.com/1051772/88618495-e672ae00-d05e-11ea-8aa2-91b2596cf3f5.jpg)

![20200727_223423](https://user-images.githubusercontent.com/1051772/88618518-f38f9d00-d05e-11ea-9bcc-3ba409c71f6d.jpg)

**Comment by GalvareyPoco** [comment]
hey, i want to do the same thing i have a k551 and it has the same chip, are you able to trace the pinout? i dont have a tester, also i have the firmware file if you think you can help me reverse it, i tried but i really have not a lot of experience with reverse engineering, i briked my keyboarrd and flashed the k556 firmware on my keyboard and it worked so the k551 and k556 must be the same in almost every aspect

**Comment by CalcProgrammer1** [comment]
I started looking into reverse engineering the K556 firmware today.  I asked Redragon support for the FW updater and they sent it to me.  I've extracted the FW and loaded it up in Ghidra and have been poking around.  I think I've found a function that defaults the flash so far.  My primary goal is to find the functions that control the LEDs so I can see if it has a direct control mode (and if not, possibly add one).

I've created a new channel on the OpenRGB discord #redragon-hacking if you want to join.  I'm posting my progress there.

https://discord.gg/Sg8tV7B

**Comment by CalcProgrammer1** [comment]
I think I've reverse engineered the matrix:

![picture](https://cdn.discordapp.com/attachments/737691756174508114/738109713610440765/unknown.png)

**Comment by kiven7e** [comment]
did you ever get direct mode to work?

**Comment by CalcProgrammer1** [comment]
Yes, check out SonixQMK: https://github.com/SonixQMK/qmk_firmware
